### ISSUE TITLE
read_stanfit_samples gives InexactError when step neq 1 in Sample

### ISSUE THREAD

**Issue by jonalm** [goedman/Stan.jl#33]
Hi Rob, I believe I've found a bug in `read_stanfit_samples`.

Minimal example:

``` julia
const simplecode = "
data {real sigma;}
parameters {real y;}
model {y ~ normal(0,sigma);}
"
stanmodel = Stanmodel(Sample(thin=5), name="simple", model=simplecode);
sim = stan(stanmodel, [Dict("sigma" => 1.)], CmdStanDir=CMDSTAN_HOME);
```

gives

```
InexactError()

 in peekchar(::IOStream) at ./iostream.jl:269
 in #skipchars#287(::Char, ::Function, ::IOStream, ::Base.UTF8proc.#isspace) at ./iostream.jl:286
 in (::Base.#kw##skipchars)(::Array{Any,1}, ::Base.#skipchars, ::IOStream, ::Function) at ./<missing>:0
 in read_stanfit_samples(::Stan.Stanmodel, ::Bool, ::Bool) at /Users/jonalm/.julia/v0.5/Stan/src/stancode.jl:382
 in #stan#16(::Bool, ::Bool, ::String, ::Function, ::Stan.Stanmodel, ::Array{Dict{String,Float64},1}, ::String) at /Users/jonalm/.julia/v0.5/Stan/src/stancode.jl:147
 in (::Stan.#kw##stan)(::Array{Any,1}, ::Stan.#stan, ::Stan.Stanmodel, ::Array{Dict{String,Float64},1}, ::String) at ./<missing>:0 (repeats 2 times)
```

I've tried to understand the `read_stanfit_samples` function, but I get lost after the sample file is opened on line 362 in stancode.jl.

**Comment by goedman** [comment]
Hi Jon

First of all my apologies that I had spelled your name wrong and thank you for the PR.

Your model will work with below change:

```
using Stan

ProjDir = dirname(@__FILE__)
cd(ProjDir) do

  const simplecode = "
  data {real sigma;}
  parameters {real y;}
  model {y ~ normal(0,sigma);}
  "

  #stanmodel = Stanmodel(Sample(thin=5), name="simple", model=simplecode);
  stanmodel = Stanmodel(thin=5, name="simple", model=simplecode);
  sim = stan(stanmodel, [Dict("sigma" => 1.)], CmdStanDir=CMDSTAN_HOME);

end
```

There is a note about this hiding somewhere in the README details section. This was a compromise between Stan and Mamba compatibility. I need to think if there is a better solution for that. Certainly I will update the wording in the hidden section is definitely not correct (preferred!).

Let me know what you think,
Rob

**Comment by goedman** [comment]
Hi Jon,

Just pushed a version that will also work with thinning inside Stan.

Regards,
Rob

**Comment by jonalm** [comment]
Great. Thanks! I have to use thin inside `Sample` to avoid storing the thinning steps into Mamba chains right? 

No worries about the name spelling, your version meant technically the same.

I have couple of comments regarding `Stan.jl`, which might improve certain aspects of the user-friendliness of your code, but feel free to disregard them completely.

First, the code moves the current path around into different directories. When the program crashes, like in the `InexactError` mentioned above, the RELP is left in the last used directory. When trying to invoke `stan` again it is easy to forget to move back to the model directory. I found myself compiling new stan models into `model_dir/tmp/tmp/tmp/tmp/tmp/tmp` etc. What about wrapping `stan` into a try-catch expression which moves back to the original directory if it crashes.

Second, it seems like the `update` and `adapt` inputs to `stan` overrides the `num_sample`, and `num_warmup` in `Sample`. It is a bit confusing with two ways of setting these parameters (and also that `adapt` is a keyword used in a different context the stan sample.csv files). I do understand that this ambiguity arises from the compromise between Stan and Mamba compatibility. What about making only one way of assigning these values (num_sample, thin and num_warmup), and a `bool` for controlling whether they apply to `cmd-stan` or the mamba output?

Another confusing point is that stan outputs 0 warmup steps, in statements like

```
Inference for Stan model: simple_model
4 chains: each with iter=(200,200,200,200); warmup=(0,0,0,0); thin=(5,5,5,5); 800 iterations saved.
```

even though stan runs 1000 warmup steps by default, it just doesn’t save them (if I have understood correctly?). But i suspect that this is a `cmd-stan` and not a `Stan.jl` issue?

Best, Jon

**Comment by jonalm** [comment]
On second thought regarding the first issue I mention, I guess the `cd(ProjDir) do` syntax you use above solves it.

**Comment by goedman** [comment]
Hi Jon,

I’ve inserted some answers. Your efforts and suggestions  to improve user friendliness are very much appreciated!

Regards,
Rob

> On Oct 28, 2016, at 04:33, Jon Alm Eriksen <notifications@github.com> wrote:
> 
> Great. Thanks! I have to use thin inside Sample to avoid storing the thinning steps into Mamba chains right?
> 
Correct. When I wrote the package (and in fact even right now) I wanted to stay compatible with Mamba’s diagnostics. Mamba stores all samples and the thinning factor in Mamba chains. Hence the 2 ways. Is storing the to be thinned out samples a problem for you?
> No worries about the name spelling, your version meant technically the same.
> 
> I have couple of comments regarding Stan.jl, which might improve certain aspects of the user-friendliness of your code, but feel free to disregard them completely.
> 
> First, the code moves the current path around into different directories. When the program crashes, like in the InexactError mentioned above, the RELP is left in the last used directory. When trying to invoke stan again it is easy to forget to move back to the model directory. I found myself compiling new stan models into model_dir/tmp/tmp/tmp/tmp/tmp/tmp etc. What about wrapping stan into a try-catch expression which moves back to the original directory if it crashes.
> 
Agreed. Tony (Kelman) suggested the cd(…) do … end and I have tried to be consistent with it since the last tagged version of Stan.jl. The downside of this construct (variables are no longer visible in REPL) occurs during testing. I usually temporarily comment out the “do" and “end" parts when testing.
> Second, it seems like the update and adapt inputs to stan overrides the num_sample, and num_warmup in Sample. It is a bit confusing with two ways of setting these parameters (and also that adapt is a keyword used in a different context the stan sample.csv files). I do understand that this ambiguity arises from the compromise between Stan and Mamba compatibility. What about making only one way of assigning these values (num_sample, thin and num_warmup), and a bool for controlling whether they apply to cmd-stan or the mamba output?
> 
Nice idea. Let me think this through.
> Another confusing point is that stan outputs 0 warmup steps, in statements like
> 
> Inference for Stan model: simple_model
> 4 chains: each with iter=(200,200,200,200); warmup=(0,0,0,0); thin=(5,5,5,5); 800 iterations saved.
> even though stan runs 1000 warmup steps by default, it just doesn’t save them (if I have understood correctly?). But i suspect that this is a cmd-stan and not a Stan.jl issue
> 
But the warmup samples should show up if you ask for them to be saved (I think). I need to check this in more detail after yesterday’s update.

I think above CmdStan output lines are intended to state what is used for inference. You’re right that it is inside CmdStan.

**Comment by jonalm** [comment]
Hi Rob,

Sorry about the delayed response, I'm travelling. Just one comment regarding saving the thinning. Is the version you just pushed saving the thinning even though the thin key word is applied to `Sample`? It would be nice to avoid saving the thinning, at least in my case I have to draw samples from a big number of input data sets, and as I know the autocorrelation length, I can reduce the stored data by a factor of 10. It is also nice to be able to pass the thinning argument to stan, as derived quantities only need sapling once each thinning cycle (I don't know if this is an issue for the `Stan.jl` implementation).

Best
Jon

**Comment by goedman** [comment]
Hi Jon,

Have a look at test_jonalm.jl in the test directory after the latest push to master. It tests 3 cases related to thinning in Stan and Stan.jl.

I believe warmup samples are never stored in the Mamba chains unless you call read_stanfit_warmup_samples().

The latest push has cleaned-up the descriptors for Mamba chains as created in Stan.jl.

Thanks for your help, confusing to mix CmdStan and Mamba!

Still thinking about your other suggestion. This is also kind of linked to a discussion inside the Stan team to redo the Stan output files.

Regards,
Rob

**Comment by jonalm** [comment]
Hi Rob,

Sorry it took me so long to get back to you. You are right regarding the warmup samples, I was just confused by the different ways of assigning the numbers. Thanks for the clarifying test script.

Feel free to close this issue.

Best
Jon

**Comment by goedman** [comment]
Hi Jon, no problem. I'll leave the issue open a bit longer. The testing showed a bug in the  stansummary program (simultaneously saving warmup samples and thinning inside Stan) crashes stansummary. I've reported this bug and Bob thinks it can be fixed before the Stan 3.0 timeframe.

Regards,
Rob

**Comment by goedman** [comment]
Daniel (core Stan team) has told me this won't be fixed until Stan 3.0. I'll close this issue for now but leave the test in place to check whenever this problem is fixed.
